File: Numerical Computing Library/NumericalLibrary-ver2/library/MatrixV1/src/Matrix.cpp
#include "../include/Matrix.hpp"
#include "../include/operations/Add.hpp"
#include "../include/operations/Subtract.hpp"
#include "../include/operations/Multiply.hpp"
#include <fstream>
#include <cmath>
#include <sstream>

// ─── Construction ─────────────────────────────────────────────────────────────

Matrix::Matrix(int r, int c) : rows(r), cols(c) {
    data.resize(r, std::vector<double>(c, 0));
}

// ─── I/O ──────────────────────────────────────────────────────────────────────

void Matrix::input() {
    std::cout << "Enter elements row-wise:\n";
    for (int i = 0; i < rows; i++)
        for (int j = 0; j < cols; j++)
            std::cin >> data[i][j];
}

void Matrix::inputFromFile(const std::string& filename) {
    std::ifstream file(filename);
    if (!file)
        throw std::runtime_error("Cannot open file: " + filename);

    file >> rows >> cols;
    data.assign(rows, std::vector<double>(cols));

    for (int i = 0; i < rows; i++)
        for (int j = 0; j < cols; j++)
            file >> data[i][j];
}

void Matrix::display() const {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++)
            std::cout << data[i][j] << " ";
        std::cout << "\n";
    }
}

void Matrix::saveToFile(const std::string& filename) const {
    std::ofstream file(filename);
    if (!file)
        throw std::runtime_error("Cannot open file for writing: " + filename);

    file << rows << " " << cols << "\n";
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++)
            file << data[i][j] << " ";
        file << "\n";
    }
}

// ─── Accessors ────────────────────────────────────────────────────────────────

int Matrix::getRows() const { return rows; }
int Matrix::getCols() const { return cols; }

double Matrix::get(int i, int j) const {
    if (i < 0 || i >= rows || j < 0 || j >= cols) {
        std::ostringstream oss;
        oss << "Matrix::get(" << i << ", " << j << ") out of range ["
            << rows << "x" << cols << "]";
        throw std::out_of_range(oss.str());
    }
    return data[i][j];
}

void Matrix::set(int i, int j, double val) {
    if (i < 0 || i >= rows || j < 0 || j >= cols) {
        std::ostringstream oss;
        oss << "Matrix::set(" << i << ", " << j << ") out of range ["
            << rows << "x" << cols << "]";
        throw std::out_of_range(oss.str());
    }
    data[i][j] = val;
}

// ─── Predicates ───────────────────────────────────────────────────────────────

bool Matrix::isSquare() const {
    return rows == cols;
}

bool Matrix::isSymmetric() const {
    if (!isSquare()) return false;
    const double tol = 1e-9;
    for (int i = 0; i < rows; i++)
        for (int j = i + 1; j < cols; j++)
            if (std::abs(data[i][j] - data[j][i]) > tol)
                return false;
    return true;
}

// ─── Utility methods ──────────────────────────────────────────────────────────

Matrix Matrix::transpose() const {
    Matrix result(cols, rows);
    for (int i = 0; i < rows; i++)
        for (int j = 0; j < cols; j++)
            result.data[j][i] = data[i][j];
    return result;
}

// --- Submatrix helper (for cofactor expansion) ---
Matrix Matrix::submatrix(int skipRow, int skipCol) const {
    Matrix result(rows - 1, cols - 1);
    int ri = 0;
    for (int i = 0; i < rows; i++) {
        if (i == skipRow) continue;
        int rj = 0;
        for (int j = 0; j < cols; j++) {
            if (j == skipCol) continue;
            result.data[ri][rj++] = data[i][j];
        }
        ri++;
    }
    return result;
}

// --- Cofactor expansion (recursive) ---
double Matrix::cofactorDet() const {
    if (rows == 1) return data[0][0];
    if (rows == 2) return data[0][0] * data[1][1] - data[0][1] * data[1][0];

    double det = 0.0;
    for (int j = 0; j < cols; j++) {
        double sign = (j % 2 == 0) ? 1.0 : -1.0;
        det += sign * data[0][j] * submatrix(0, j).cofactorDet();
    }
    return det;
}

// --- determinant() ---
double Matrix::determinant(DetMethod method) const {
    if (!isSquare())
        throw std::runtime_error("determinant() requires a square matrix");

    if (method == DetMethod::Cofactor)
        return cofactorDet();

    // LU-based: partial pivoting, track swap count for sign
    int n = rows;
    Matrix A(*this);
    int swaps = 0;

    for (int i = 0; i < n; i++) {
        // Find pivot
        int pivot = i;
        for (int p = i + 1; p < n; p++)
            if (std::abs(A.data[p][i]) > std::abs(A.data[pivot][i]))
                pivot = p;

        if (std::abs(A.data[pivot][i]) < 1e-12)
            return 0.0;  // singular — determinant is zero

        if (pivot != i) {
            std::swap(A.data[i], A.data[pivot]);
            swaps++;
        }

        for (int j = i + 1; j < n; j++) {
            double factor = A.data[j][i] / A.data[i][i];
            for (int k = i; k < n; k++)
                A.data[j][k] -= factor * A.data[i][k];
        }
    }

    double det = (swaps % 2 == 0) ? 1.0 : -1.0;
    for (int i = 0; i < n; i++)
        det *= A.data[i][i];
    return det;
}

// --- inverse() — Gauss-Jordan elimination on [A | I] ---
Matrix Matrix::inverse() const {
    if (!isSquare())
        throw std::runtime_error("inverse() requires a square matrix");

    int n = rows;

    // Build augmented matrix [A | I]
    Matrix aug(n, 2 * n);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++)
            aug.data[i][j] = data[i][j];
        aug.data[i][n + i] = 1.0;  // identity block
    }

    // Forward elimination with partial pivoting
    for (int i = 0; i < n; i++) {
        // Find pivot
        int pivot = i;
        for (int p = i + 1; p < n; p++)
            if (std::abs(aug.data[p][i]) > std::abs(aug.data[pivot][i]))
                pivot = p;

        if (std::abs(aug.data[pivot][i]) < 1e-12)
            throw std::runtime_error("Matrix is singular — inverse does not exist");

        if (pivot != i)
            std::swap(aug.data[i], aug.data[pivot]);

        // Scale pivot row
        double scale = aug.data[i][i];
        for (int j = 0; j < 2 * n; j++)
            aug.data[i][j] /= scale;

        // Eliminate column above and below
        for (int k = 0; k < n; k++) {
            if (k == i) continue;
            double factor = aug.data[k][i];
            for (int j = 0; j < 2 * n; j++)
                aug.data[k][j] -= factor * aug.data[i][j];
        }
    }

    // Extract right block as the inverse
    Matrix result(n, n);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            result.data[i][j] = aug.data[i][n + j];

    return result;
}

// ─── Arithmetic operators ─────────────────────────────────────────────────────

Matrix operator+(const Matrix& A, const Matrix& B) { return Add::compute(A, B); }
Matrix operator-(const Matrix& A, const Matrix& B) { return Subtract::compute(A, B); }
Matrix operator*(const Matrix& A, const Matrix& B) { return Multiply::compute(A, B); }

// ─── Comparison operators ─────────────────────────────────────────────────────

bool operator==(const Matrix& A, const Matrix& B) {
    if (A.rows != B.rows || A.cols != B.cols) return false;
    const double tol = 1e-9;
    for (int i = 0; i < A.rows; i++)
        for (int j = 0; j < A.cols; j++)
            if (std::abs(A.data[i][j] - B.data[i][j]) > tol)
                return false;
    return true;
}

bool operator!=(const Matrix& A, const Matrix& B) { return !(A == B); }

// ─── Stream output ────────────────────────────────────────────────────────────

std::ostream& operator<<(std::ostream& os, const Matrix& M) {
    for (int i = 0; i < M.rows; i++) {
        for (int j = 0; j < M.cols; j++) {
            os << M.data[i][j];
            if (j < M.cols - 1) os << " ";
        }
        os << "\n";
    }
    return os;
}

File: Numerical Computing Library/NumericalLibrary-ver2/library/MatrixV1_Python/matrixv1/core/matrix.py
"""Core Matrix class — mirrors MatrixV1 C++ Matrix interface."""

from __future__ import annotations

import math
import copy
from typing import List


class Matrix:
    """2-D matrix of floats backed by a list-of-lists.

    Indexing is 0-based throughout. All mutating operations return a new
    Matrix; the original is never modified.
    """

    # ── Construction ─────────────────────────────────────────────────────────

    def __init__(self, rows: int = 0, cols: int = 0) -> None:
        if rows < 0 or cols < 0:
            raise ValueError(f"Dimensions must be non-negative, got ({rows}, {cols})")
        self._rows = rows
        self._cols = cols
        self._data: List[List[float]] = [
            [0.0] * cols for _ in range(rows)
        ]

    @classmethod
    def from_list(cls, data: List[List[float]]) -> "Matrix":
        """Build a Matrix from a nested list (no copy of outer list, but rows are copied)."""
        if not data:
            return cls(0, 0)
        rows = len(data)
        cols = len(data[0])
        m = cls(rows, cols)
        for i, row in enumerate(data):
            if len(row) != cols:
                raise ValueError("All rows must have the same length")
            for j, val in enumerate(row):
                m._data[i][j] = float(val)
        return m

    @classmethod
    def from_file(cls, filename: str) -> "Matrix":
        """Load from file. Format: first line is 'rows cols', then data row-by-row."""
        with open(filename) as f:
            rows, cols = map(int, f.readline().split())
            m = cls(rows, cols)
            for i in range(rows):
                values = list(map(float, f.readline().split()))
                for j, v in enumerate(values):
                    m._data[i][j] = v
        return m

    # ── Accessors ─────────────────────────────────────────────────────────────

    @property
    def rows(self) -> int:
        return self._rows

    @property
    def cols(self) -> int:
        return self._cols

    def _check_bounds(self, i: int, j: int) -> None:
        if i < 0 or i >= self._rows or j < 0 or j >= self._cols:
            raise IndexError(
                f"Matrix index ({i}, {j}) out of range [{self._rows}x{self._cols}]"
            )

    def get(self, i: int, j: int) -> float:
        self._check_bounds(i, j)
        return self._data[i][j]

    def set(self, i: int, j: int, val: float) -> None:
        self._check_bounds(i, j)
        self._data[i][j] = float(val)

    # ── Predicates ────────────────────────────────────────────────────────────

    def is_square(self) -> bool:
        return self._rows == self._cols

    def is_symmetric(self, tol: float = 1e-9) -> bool:
        if not self.is_square():
            return False
        for i in range(self._rows):
            for j in range(i + 1, self._cols):
                if abs(self._data[i][j] - self._data[j][i]) > tol:
                    return False
        return True

    # ── Utilities ─────────────────────────────────────────────────────────────

    def transpose(self) -> "Matrix":
        result = Matrix(self._cols, self._rows)
        for i in range(self._rows):
            for j in range(self._cols):
                result._data[j][i] = self._data[i][j]
        return result

    def _submatrix(self, skip_row: int, skip_col: int) -> "Matrix":
        """Return (n-1)×(n-1) matrix with row skip_row and col skip_col removed."""
        result = Matrix(self._rows - 1, self._cols - 1)
        ri = 0
        for i in range(self._rows):
            if i == skip_row:
                continue
            rj = 0
            for j in range(self._cols):
                if j == skip_col:
                    continue
                result._data[ri][rj] = self._data[i][j]
                rj += 1
            ri += 1
        return result

    def _cofactor_det(self) -> float:
        """Recursive cofactor expansion — O(n!), for small matrices / educational use."""
        if self._rows == 1:
            return self._data[0][0]
        if self._rows == 2:
            return (self._data[0][0] * self._data[1][1]
                    - self._data[0][1] * self._data[1][0])
        det = 0.0
        for j in range(self._cols):
            sign = 1.0 if j % 2 == 0 else -1.0
            det += sign * self._data[0][j] * self._submatrix(0, j)._cofactor_det()
        return det

    def determinant(self, method: str = "lu") -> float:
        """Compute determinant via 'lu' (default, O(n³)) or 'cofactor' (O(n!))."""
        if not self.is_square():
            raise ValueError("determinant() requires a square matrix")
        if method == "cofactor":
            return self._cofactor_det()
        # LU-based: partial pivoting, count swaps for sign
        n = self._rows
        A = copy.deepcopy(self._data)
        swaps = 0
        for i in range(n):
            pivot = max(range(i, n), key=lambda p: abs(A[p][i]))
            if abs(A[pivot][i]) < 1e-12:
                return 0.0
            if pivot != i:
                A[i], A[pivot] = A[pivot], A[i]
                swaps += 1
            for j in range(i + 1, n):
                factor = A[j][i] / A[i][i]
                for k in range(i, n):
                    A[j][k] -= factor * A[i][k]
        det = -1.0 if swaps % 2 else 1.0
        for i in range(n):
            det *= A[i][i]
        return det

    def inverse(self) -> "Matrix":
        """Gauss-Jordan elimination on [A | I]. Raises ValueError if singular or non-square."""
        if not self.is_square():
            raise ValueError("inverse() requires a square matrix")
        n = self._rows
        # Build augmented [A | I]
        aug = [[0.0] * (2 * n) for _ in range(n)]
        for i in range(n):
            for j in range(n):
                aug[i][j] = self._data[i][j]
            aug[i][n + i] = 1.0
        # Forward elimination with partial pivoting
        for i in range(n):
            pivot = max(range(i, n), key=lambda p: abs(aug[p][i]))
            if abs(aug[pivot][i]) < 1e-12:
                raise ValueError("Matrix is singular — inverse does not exist")
            if pivot != i:
                aug[i], aug[pivot] = aug[pivot], aug[i]
            scale = aug[i][i]
            for j in range(2 * n):
                aug[i][j] /= scale
            for k in range(n):
                if k == i:
                    continue
                factor = aug[k][i]
                for j in range(2 * n):
                    aug[k][j] -= factor * aug[i][j]
        result = Matrix(n, n)
        for i in range(n):
            for j in range(n):
                result._data[i][j] = aug[i][n + j]
        return result

    # ── I/O ───────────────────────────────────────────────────────────────────

    def to_file(self, filename: str) -> None:
        with open(filename, "w") as f:
            f.write(f"{self._rows} {self._cols}\n")
            for row in self._data:
                f.write(" ".join(str(v) for v in row) + "\n")

    def input_from_console(self) -> None:
        print(f"  Enter {self._rows * self._cols} elements row-wise:")
        for i in range(self._rows):
            for j in range(self._cols):
                self._data[i][j] = float(input(f"  [{i}][{j}]: "))

    # ── Operators ─────────────────────────────────────────────────────────────

    def __add__(self, other: "Matrix") -> "Matrix":
        from matrixv1.operations.add import add
        return add(self, other)

    def __sub__(self, other: "Matrix") -> "Matrix":
        from matrixv1.operations.subtract import subtract
        return subtract(self, other)

    def __mul__(self, other: "Matrix") -> "Matrix":
        from matrixv1.operations.multiply import multiply
        return multiply(self, other)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Matrix):
            return NotImplemented
        if self._rows != other._rows or self._cols != other._cols:
            return False
        tol = 1e-9
        for i in range(self._rows):
            for j in range(self._cols):
                if abs(self._data[i][j] - other._data[i][j]) > tol:
                    return False
        return True

    def __ne__(self, other: object) -> bool:
        result = self.__eq__(other)
        if result is NotImplemented:
            return result
        return not result

    def __repr__(self) -> str:
        return f"Matrix({self._rows}x{self._cols})"

    def __str__(self) -> str:
        lines = []
        for row in self._data:
            lines.append("  " + "  ".join(f"{v:10.6g}" for v in row))
        return "\n".join(lines)

    def display(self) -> None:
        print(str(self))

    # ── Internal helpers (used by solvers) ────────────────────────────────────

    def _raw(self) -> List[List[float]]:
        """Return a deep copy of the internal data list."""
        return copy.deepcopy(self._data)
